src/NumberInput/NumberInput.svelte

```diff
@@ -137,16 +141,67 @@
     $$props["aria-label"] ||
     "Numeric input field with increment and decrement buttons";
 
+  //$: value && !!changeCaretPosition;
+
   function parse(raw) {
     return raw != "" ? Number(raw) : null;
   }
 
-  function onInput({ target }) {
+  function updatePreviousValue() {
+    previousValue = parse(value);
+  }
+
+  async function onInput({ target, inputType, data }) {
+    const possibleNewValue = parse(target.value);
     value = parse(target.value);
 
+    const isIntegerTransition =
+      !data &&
+      !Number.isInteger(previousValue) &&
+      Number.isInteger(possibleNewValue);
+
+    if (isIntegerTransition && value < previousValue) {
+      await setSelectionAfterTick(target.value.length);
+    } else if (isIntegerTransition && previousValue && value > previousValue) {
+      let parts = previousValue.toString().split(".");
+
+      if (inputType === "deleteContentForward" && parts[1]) {
+        parts[1] = parts[1].substring(1);
+      } else if (inputType === "deleteContentBackward") {
+        parts[0] = parts[0].slice(0, -1);
+      }
+
+      const chars =
+        parts[0].length + (inputType === "deleteContentBackward" ? 1 : 0);
+      value = parse(parts.join("."));
+      await setSelectionAfterTick(chars);
+    } else if (
+      !isIntegerTransition &&
+      previousValue &&
+      value === previousValue &&
+      inputType === "deleteContentForward"
+    ) {
+      value = parse(0);
+      await setSelectionAfterTick(0);
+    } else {
+      value = possibleNewValue;
+    }
```

Comment:
This entire logic is hard to understand/follow, though I also don't know if/how it could be improved...